clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -500,14 +500,22 @@ boolean joinGroupIfNeeded(final Timer timer) {
                     requestRejoin(shortReason, fullReason);
                 }
 
+                // continue to retry as long as the timer hasn't expired
```

Comment:
Could we simplify this multi-if logic as:

```
if (!future.isRetriable()) { throw }
else {
    if (timer.isExpired() { return false }
    else if (exception instance of.. ) { continue}
    else {timer.sleep(..)}
}
```

Also could we add a comment on top clarifying that the order of precedence are deliberated in this order and future changes should pay attention to not change it unnecessarily.

Reply:
I think the `instanceof ...` exceptions are also non-retriable, and I think they need to be handled first.

Reply:
so the if else blocks becomes a bit fragmented.  or we could do (imo, a bit more difficult to read):
```
if (!future.isRetriable()) {
  if ( ... instance of ... ) { continue; }
  throw ...
}

{rest of the logic there}
```

Reply:
However, this is a bit more nested, which can be harder to read